torch/csrc/distributed/rpc/tensorpipe_agent.cpp

```diff
@@ -0,0 +1,367 @@
+#include <caffe2/torch/csrc/distributed/rpc/tensorpipe_agent.h>
+
+#include <torch/csrc/distributed/rpc/request_callback_impl.h>
+#include <torch/csrc/distributed/rpc/utils.h>
+
+#include <tensorpipe/channel/basic/basic.h>
+#include <tensorpipe/channel/cma/cma.h>
+#include <tensorpipe/transport/shm/context.h>
+#include <tensorpipe/transport/uv/context.h>
+
+namespace torch {
+namespace distributed {
+namespace rpc {
+
+using clock = std::chrono::high_resolution_clock;
+using nss = std::chrono::nanoseconds;
+
+std::atomic<uint32_t> TensorPipeAgent::nextShmAddressID_(0);
+
+TensorPipeAgent::TensorPipeAgent(
+    worker_id_t selfId,
+    std::string selfName,
+    std::shared_ptr<::c10d::Store> addressStore,
+    TensorPipeRpcBackendOptions opts)
+    : RpcAgent(
+          WorkerInfo(selfName, selfId),
+          std::make_unique<RequestCallbackImpl>(),
+          opts.rpcTimeout),
+      context_(tensorpipe::Context::create()),
+      addressStore_(std::move(addressStore)),
+      opts_(std::move(opts)) {
+  // Generate the maps for once.
+  for (const std::pair<std::string, worker_id_t>& kv : opts_.workerNameToId) {
+    const string& workerName = kv.first;
+    worker_id_t workerId = kv.second;
+    workerIdToInfo_.insert({workerId, WorkerInfo(workerName, workerId)});
+    workerNameToInfo_.insert({workerName, WorkerInfo(workerName, workerId)});
+  }
+}
+
+TensorPipeAgent::~TensorPipeAgent() {
+  shutdown();
+}
+
+void TensorPipeAgent::start() {
+  context_->registerTransport(
+      1, "tcp", std::make_shared<tensorpipe::transport::uv::Context>());
+  context_->registerTransport(
+      0, "shm", std::make_shared<tensorpipe::transport::shm::Context>());
+  context_->registerChannelFactory(
+      1,
+      "basic",
+      std::make_shared<tensorpipe::channel::basic::BasicChannelFactory>());
+  context_->registerChannelFactory(
+      0, "cma", tensorpipe::channel::cma::CmaChannelFactory::create());
+
+  std::vector<std::string> addresses = {createUniqueShmAddr(),
+                                        "tcp://127.0.0.1"};
```

Comment:
Ideally tensorpipe could provide a helper to get IP address for given device interface or host names, or return the IP address of the default host name. More specifically like Gloo https://github.com/facebookincubator/gloo/blob/master/gloo/transport/tcp/device.cc#L142. As this is not supported in tensorpipe right now, we could have a couple of work around:
* Retrieve the default IP address on rpc agent layer
* Let user specify IP address by rpcBackendOption
* Hard code 127.0.0.1 for now for single machine only